feat(gatsby-plugin-fastify): add customizable headers with sensible defaults for security

tsdexter opened the pull request:
## Description

@moonmeister I think I've got this working as expected now. Please take a look and let me know. 

- `options.headers` plugin configuration added for applying custom headers to specific files/paths
- default security headers added

### Documentation

Headers section at the bottom of the README

## Related Issues

closes #98 
fixes #59 

@moonmeister test `Gatsby Node API › Should Build Config` is failing - I'm not sure why. Can you take a look?

moonmeister commented on packages/gatsby-plugin-fastify/src/utils/headers.ts:
```diff
@@ -38,3 +47,34 @@ export function moduleHeaderDecorator(this: FastifyReply, module: Modules): void
+  let routeHeaders = {};
+  // console.log(`\n\nheaders`, {headers}, `\n\n`);
+  mapObject(headers, (pattern, headerObject) => {
+    if (isMatch(route, pattern)) {
```
We should include match examples in docs (I believe you did this), but should also include a note that we use `picomatch` and link to those complete docs.

moonmeister replied:
Dope

moonmeister commented on packages/gatsby-plugin-fastify/src/utils/headers.ts:
```diff
@@ -38,3 +47,34 @@ export function moduleHeaderDecorator(this: FastifyReply, module: Modules): void
+   * we expect this but we need a way to prioritize one match over another
+   * or am I over complicating this?
+   *
+   * @tsdexter: I think we can just use the order of the headers in the
```
Works for me, Let's clean up this comment with a clear description of how we expect this to function.

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/server-routes.ts:
```diff
@@ -107,6 +109,13 @@ export const handleServerRoutes: FastifyPluginAsync<{
             const results = await renderHTML({ data: pageQueryData });
 
+            if (page.mode === "DSG") {
+              mapValues(NEVER_CACHE_HEADER, (value, key) => {
```
Can this not use the `appendRouteHeaders` function?

tsdexter replied:
I did remove it initially and just had the `appendRouteHeaders` call below it but then realized `NEVER_CACHE_HEADERS` was not a part of `appendRouteHeaders` but DSG pages needed it, so I added it back in. I think this will change as I refactor and clean it up.

moonmeister reviewed (changes requested):
Good start! Thanks for all your work on this. I'm sure a lot of my comments are against the code you copied and pasted from me, I'm aware. None of this is a comment on you, just me wanting to get us to a good point. 

So the major question in the back of my head is if this can/should be implemented as a plugin using [hooks](https://www.fastify.io/docs/latest/Reference/Hooks/) or [Decorators](https://www.fastify.io/docs/latest/Reference/Decorators/). I myself have been refactoring features as I go to use these features. It helps segment code logic into a single file so I don't have to reimplement things across ssr/static, etc.

Looks like your tests have focussed on static/ssr/dsg. Have you tested client-only routes or thought about how headers interact with client-only routes?

I'm sure I've forgotten some things but hopefully this gets you started?

moonmeister commented:
> @moonmeister what's the best way to test this out in a real site pre-release? For example, can we do an `experimental` release?

Great question, It is possible, but I've never done it with Changesets, I'll try to make time to do some reading...though if you want to look into the docs and see what's possible that'd be helpful.

moonmeister commented:
> > @moonmeister what's the best way to test this out in a real site pre-release? For example, can we do an `experimental` release?
> 
> Great question, It is possible, but I've never done it with Changesets, I'll try to make time to do some reading...though if you want to look into the docs and see what's possible that'd be helpful.

I think I'll need to manually do a pre-release. LEt's get this a little further along then I can do that.

tsdexter commented:
Thanks @moonmeister - getting back to it today. 

FYI:
 
> @moonmeister test `Gatsby Node API › Should Build Config` is failing - I'm not sure why. Can you take a look?

I think I figured out why this is failing, looking at a fix. Having to do with trying to read the `webpack.stats.json` from the `__tests__` files which is now used in `onPostBuild` but wasn't previously

tsdexter commented on packages/gatsby-plugin-fastify/src/gatsby-node.ts:
```diff
@@ -62,6 +89,9 @@ export const pluginOptionsSchema: GatsbyNode["pluginOptionsSchema"] = ({ Joi })
           return value;
         }),
     }).default(),
+    headers: headersSchema,
+    mergeSecurityHeaders: Joi.boolean().default(true),
+    mergeCacheHeaders: Joi.boolean().default(true),
```
@moonmeister perhaps a `use` or `useStrategy` naming convention, ie: `useGatsbyCaching`/`useGatsbyCachingStrategy` with a link to: https://www.gatsbyjs.com/docs/how-to/previews-deploys-hosting/caching/ and `useSecurity`/`useSecurityStrategy` with a link to our docs/defaults?

tsdexter replied:
I don't think so. Shorter than the current include option!

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/static.ts:
```diff
@@ -17,14 +17,9 @@ export const handleStatic: FastifyPluginAsync<Partial<FastifyStaticOptions>> = f
-        } else {
-          reply.setHeader(...NEVER_CACHE_HEADER);
-        }
+        const barePath = path.replace(publicPath, ``).replace(/\\/g, `/`);
```
@moonmeister I'm guessing it's a windows only issue, but my paths were like `C:\Documents\path\to\public\path\to\route` vs the Fastify path like `/path/to/route` so I needed to replace the `\`'s with `/`'s. I'll be sure to test it in OSX as well and comment it.

moonmeister replied:
Very possibly, there are built in node path functions for this that can help. `path.posix` I believe.

tsdexter commented on packages/gatsby-plugin-fastify/src/utils/constants.ts:
```diff
@@ -3,16 +3,31 @@ export const PATH_TO_FUNCTIONS = ".cache/functions";
+export const IMMUTABLE_CACHING_HEADER = {
+  "cache-control": `public, max-age=31536000, immutable`,
+};
+export const NEVER_CACHE_HEADER = { "cache-control": `public, max-age=0, must-revalidate` };
```
@moonmeister Did you mean sticking with objects as it is currently? I'm not sure which is better, I believe I stuck to the object because the existing code had it both ways and it was simpler to convert it all to objects than vice versa. I will re-visit as I refactor.

moonmeister replied:
I don't care which way it falls, let's just be consistent.

tsdexter commented:
@moonmeister I've refactored the code with a new headers plugin that takes care of adding all of the necessary headers for all modules in one place via hooks/decorators. It now also supports the rest of the modules as well. I've also moved tests for headers out of each modules test suite into a new headers specific test suite and added test coverage for `header-builder.ts`. I think it should resolve all of the previous comments. Please take a look and let me know!

moonmeister commented:
Already looking better, but I realized I need to do this on a computer. Will try and get to it soon. Thanks @tsdexter !!!!

moonmeister commented:
Thanks again @tsdexter ...we're getting closer...step by step. Lots in that review...sorry. I can make some time to write code soon so if you are tight on time, do what you can and let me know specific things I can help with...but do as much as you want. I'm happy to keep providing feedback.

tsdexter commented on packages/gatsby-plugin-fastify/src/gatsby/header-builder.ts:
```diff
@@ -0,0 +1,85 @@
+
+function deepMerge(...headers: Headers[]) {
+  // eslint-disable-next-line unicorn/no-array-reduce
+  return headers.reduce((accumulator, header) => {
```
resolved in 67e1817fd1b2eac5241cfbfe0fe3999e45c09174

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/decorators.ts:
```diff
@@ -3,15 +3,19 @@ import fp from "fastify-plugin";
 export const implementUtilDecorators: FastifyPluginAsync = fp(async (fastify) => {
   fastify.decorateReply("setHeader", setHeaderDecorator);
   fastify.decorateReply("appendModuleHeader", moduleHeaderDecorator, ["setHeader"]);
+  // eslint-disable-next-line unicorn/no-null
```
I set it to `null` as suggested in [fastify docs](https://www.fastify.io/docs/latest/Reference/Decorators/): 

> Note that it is important to keep the initial shape of a decorated field as close as possible to the value intended to be set dynamically in the future. Initialize a decorator as a `''` if the intended value is a string, and as `null` if it will be an object or a function.

I'm not sure the intention/benefits of the linting rule, but I assumed to go with the vendor suggestion. Which way do you want to land?

tsdexter replied:
I changed this around a bit in 9fcd9b1dcd51196968efb43ab4d923599a98cd20.

Since the type is now `ISSRData["serverDataHeaders"]` which includes `| undefined` I tried switching the initialization to use `undefined` but the `lint-staged` ended up rewriting it to `fastify.decorateReply("serverDataHeaders");` which then causes a `too few arguments` issue so I amended it back to using `null`. I think we need to keep it `null` but let me know if you think otherwise.

moonmeister replied:
Yeah, tell should be fine. Rules are meant to be broken. https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-null.md has some info on the reasoning if you're curious.

tsdexter commented on packages/gatsby-plugin-fastify/package.json:
```diff
@@ -53,9 +59,10 @@
-    "babel-jest": "^29.4.3",
-    "babel-preset-gatsby-package": "^3.7.0",
+    "@types/webpack-assets-manifest": "^5.1.0",
+    "@types/yargs": "^17.0.18",
+    "babel-jest": "^29.3.1",
+    "babel-preset-gatsby-package": "^3.3.0",
```
updated in: 42005d250b9ac8cb56c4abde762fb58137e020ad

tsdexter commented on packages/gatsby-plugin-fastify/src/__tests__/gatsby-node.js:
```diff
@@ -1,5 +1,6 @@
-import { onPostBuild } from "../gatsby-node";
 import fs from "fs-extra";
+import mockPath from "node:path";
```
Importing as `path` causes the error: 

>The module factory of `jest.mock()` is not allowed to reference any out-of-scope variables.

Importing as `mockPath` fixes the scope. I took this from: https://lukerogerson.medium.com/two-ways-to-fix-the-jest-test-error-the-module-factory-of-jest-mock-is-not-allowed-to-bf022b5175dd

tsdexter commented on packages/gatsby-plugin-fastify/src/gatsby-node.ts:
```diff
@@ -61,6 +88,11 @@ export const pluginOptionsSchema: GatsbyNode["pluginOptionsSchema"] = ({ Joi })
           return value;
         }),
+      headers: Joi.object({
+        customHeaders: headersSchema,
```
resolved in: 0b31e919048e0bbc8a186903a5f85b25f8895a12

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/gatsby.ts:
```diff
@@ -14,6 +14,7 @@ import fastifyAccepts from "@fastify/accepts";
 import middiePlugin from "@fastify/middie";
 
 import type { FastifyPluginAsync } from "fastify";
+import { handleHeaders } from "./headers";
```
resolved in 9fcd9b1dcd51196968efb43ab4d923599a98cd20

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,44 @@
+import { isMatch } from "picomatch";
+import type { FastifyPluginAsync, FastifyReply } from "fastify";
+import { getConfig } from "../utils/config";
+import { NEVER_CACHE_HEADER } from "../utils/constants";
```
resolved in 9fcd9b1dcd51196968efb43ab4d923599a98cd20

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,44 @@
+          headers: { useDefaultCaching },
+        },
+      },
+    } = getConfig();
```
resolved in 9fcd9b1dcd51196968efb43ab4d923599a98cd20

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/server-routes.ts:
```diff
@@ -57,19 +57,15 @@ export const handleServerRoutes: FastifyPluginAsync<{
 
-          const pageData = (await renderPageData({ data: pageQueryData })) as unknown;
+          // add pageQueryData to reply so it can be accessed in the onSend hook
+          reply.pageQueryData = pageQueryData;
```
It's not within the SSR specific, rather the HTML/page-data specific areas as you mentioned in the second comment, hence running it twice.

`page-data` should also receive the same headers as `.html` as mentioned in the [gatsby caching docs](https://www.gatsbyjs.com/docs/how-to/previews-deploys-hosting/caching/#page-data) which is how it is working. Both SSR page and page-data have caching and security headers set, if defaults are `true` in the config, via: `SECURITY_HEADERS` and `CACHING_HEADERS` from `constants.ts`, this part only handles overwriting them if they are set within `getServerData`, it doesn't handle the defaults.

tsdexter replied:
I checked the gatsby hosted SSR as well as building a new from this branch. The SSR `cache-control` is indeed the same across `gserve` and gatsby hosting: 
<img width="664" alt="image" src="https://user-images.githubusercontent.com/4126987/225034539-47a02822-3eb8-4602-bdfc-63dac3ca44b8.png">
<img width="320" alt="image" src="https://user-images.githubusercontent.com/4126987/225035191-0f512c30-c5b4-4a10-937e-286b88e4cdd7.png">

moonmeister replied:
Thanks for confirming all this.

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/server-routes.ts:
```diff
@@ -105,20 +101,13 @@ export const handleServerRoutes: FastifyPluginAsync<{
-            }
+            const results = await renderHTML({ data: pageQueryData });
 
-            if (page.mode === "DSG") {
```
This may have been a mistake previously as gatsby hosting does set the `NEVER_CACHE_HEADER` on SSR routes as [seen above](https://github.com/gatsby-uc/plugins/pull/392#discussion_r1135656503)

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,44 @@
+    } = getConfig();
+
+    // get the url without query params
+    const url = request.raw.url?.split(`?`)[0];
```
Yes, `request.url` does indeed work. I actually noticed you've already done this elsewhere with a utility `removeQueryParamsFromUrl()` so I've switched it to reuse that instead. Resolved in: 9fcd9b1dcd51196968efb43ab4d923599a98cd20

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,44 @@
+        (!url?.includes(".") && !url?.endsWith("/"))) // no trailing slash or file extension, so it's a directory
+    ) {
+      reply.headers(NEVER_CACHE_HEADER);
+    }
```
>What would would this NOT apply to? 

It shouldn't apply to any specific static files (images, js, css, pdfs, docs, etc) just `directories` either with/without a trailing slash, where the actual served file is the `/index.html` so it should only apply `NEVER_CACHE_HEADER` to `.html` files (and function routes: see below) which matches gatsby docs/hosting. It could be missing something, but I think it's right. 

> I'm also wondering if it is still to broad... this shouldn't apply to functions

You are correct in that this will apply to function routes like `/api/test/`. I've updated the `plugin-fastify` test site to support running functions on gatsby hosting in 9a328fb662ff04347d63d235798f3bef1b538eb4 and verified that gatsby hosting [does indeed](https://pluginsfeatfastifydefaultcusto.gatsbyjs.io/api/test) use the `NEVER_CACHE_HEADER` for function routes as well. 

![image](https://user-images.githubusercontent.com/4126987/225116459-96e0b600-307a-4edb-9318-0e70b2cd7268.png)

> Maybe we need to tag routes with their type e.g. like the way x-gatsby-server does but structure the data so that its SSR|| DSG || STATIC || FUNCTIONS etc...

In the first refactor I did go down the path of tagging routes with their module by adding a decorator `reply.module` which I was going to then use in the headers plugin to set the `FG_MODULE_HEADER` based on the module used as well. I had this working for all modules except `static.ts` because it is globally scoped, since the `fastify-static` plugin doesn't have any hooks to run when it's actually serving a file the `Static` module was added to all replies, not just the ones actually served by static.

I tried quite a few methods such as narrowing the scope of static (however 500/404/client-routes needed `reply.sendFile`), adding separate `fastify-static` implementations within 500/404/client, adding the `reply.module` within `fastify-static`'s `setHeaders` option, and some others I can't recall now...I was not able to reliably only add the static module to files actually served by `fastify-static` though I may have been missing something obvious? 

> Also, where is the immutable cache header getting applied?

See dupe question below which is more specific: https://github.com/gatsby-uc/plugins/pull/392#discussion_r1134965306

tsdexter replied:
resolved in 897e7d519c1752f8ce150f1d583dcd8728b98ab2

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/static.ts:
```diff
@@ -16,15 +15,7 @@ export const handleStatic: FastifyPluginAsync<Partial<FastifyStaticOptions>> = f
       redirect: false,
       wildcard: false,
-      setHeaders: (reply, path) => {
-        if (
-          isMatch(path, ["**/public/*.@(js|css)", "**/public/static/**"]) &&
-          isMatch(path, "!**/sw.js")
-        ) {
-          reply.setHeader(...IMMUTABLE_CACHING_HEADER);
```
The `IMMUTABLE_CACHING_HEADER` is added in two spots.

In `constants.ts`, it is added to `/static/**` this is equivalent to the above `**/public/static/**` since the pattern matching will now be done on the `request.url` vs. the static file path that `fastify-static` passes in. 

```
export const CACHING_HEADERS = {
  "/static/**": IMMUTABLE_CACHING_HEADER,
```

As well, in `header-builder.ts` it is added to all chunked files consistent with gatsby hosting. 

I'm not sure where this rule came from `"**/public/*.@(js|css)"`. I believe only chunked files should be `immutable`, not all `.js`, `.css` file within the root. I noticed this on my own site where generated files in the root for lazy loaded components do not get a chunk entry and gatsby hosting does not set any `cache-control` headers at all for them. 

In order to reproduce and test this behaviour, I added a lazy loaded component as well as support for "unsetting" headers in 8fe0bc0fe4a595d59f481e25fdb47e343c5cd619 which results in both gatsby hosting and the fastify server using the same "unset" `cache-control` header for these `.js` files as opposed to an immutable header:
![image](https://user-images.githubusercontent.com/4126987/225120112-636d5b72-9e83-40f1-ba1b-024288d87fbc.png)

For `.css` files in the root, I'm not sure how to reproduce actually including a CSS file from the root as importing any CSS gets added to the CSS chunk and served as a chunked file which gets the appropriate immutable header due to being chunked.

moonmeister replied:
Ooooookay...I see what's happening. Thanks for the explanation.  I think this reinforces my desire for this to all happen at build time. Right now it seems that the application of default and custom headers is split between build-time (`gatsby/header-build.ts`), and runtime (`plugins/headers.ts). From a code perspective, this seems confusing, not to mention the potential performance issues. 

So, the solution is to enumerate all possible routes at build time (already done)...and use those to calculate headers for that route. This can be written to the config and used at runtime to actually apply those headers to each route in `plugins/headers.ts`.

This generally applies to all route paterns `/example/[post_slug].ts`. The major exception being if the header applies to `/example/test/`. We don't know that that exists but we will assume it does and save the header in the lookup table. That may take some special logic. lmk if you have any questions.

As for CSS files, nice catch on the CSS and JS files. IIRC that was an intentional simplification, though technically a bug. If I'm not wrong, the only JS and CSS files NOT hashed would be those included in the `static` folder. Are lazy loaded components also an exception? Atleast for CSS, the way to test would be adding a test fil in `src/static`, this will get coppied to `public` without a hash I believe.

tsdexter replied:
Resolved in ab6ff63f7c327471f681e9c067ea236a58f52163

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,44 @@
+
+    // handle pattern match caching
+    for (const [pattern, headerObject] of Object.entries(headers)) {
+      if (url && isMatch(url, pattern)) {
```
Good catch, we don't need to loop if no URL. Resolved in 9fcd9b1dcd51196968efb43ab4d923599a98cd20

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,44 @@
+        reply.headers(headerObject);
+
+        // overwrite configured headers with SSR values
+        if (reply.pageQueryData?.serverDataHeaders) {
```
`reply.headers` doesn't erase all headers and then set new ones. It only sets/resets the ones passed in so in this case, it will only overwrite previously set headers that are set in `serverDataHeaders`. I've updated the comment to reflect that in  9fcd9b1dcd51196968efb43ab4d923599a98cd20 

`// allow serverDataHeaders to override duplicate configured headers`

tsdexter commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,44 @@
+        }
+      }
+    }
+  });
```
I definitely see how this might be a problem though I personally like Option 3 😉😏

Joking aside, if option 3 isn't feasible, I would lean toward build time as well. I'll see about running the `benchmark.js` file tomorrow morning and compare (unless you have time during review tonight).

moonmeister replied:
Let's start with bench marks (I won't have time) and go from there. That script runs on the current branch. So recomendation would be to run on main a few times then checout this branch and run again. The more things you can close on your computer the more accurate the results will be. Basically close everything but a raw terminal and execute from there.

moonmeister replied:
Let's start with bench marks (I won't have time) and go from there. That script runs on the current branch. So recomendation would be to run on main a few times then checout this branch and run again. The more things you can close on your computer the more accurate the results will be. Basically close everything but a raw terminal and execute from there.

moonmeister commented:
Okay, so I think we need to document some of these complex interactions of what does and doesn't recieve various headers. This will help us. future maintainers, and users hopefully. PLEASE VERIFY EVERYTHING HERE. Things with question marks mean I didn't know...Answers with an additional question mark mean I was unsure. Once we verify all this and finalize all answers we can merge. Please add any rows/columns I may habve missed. 

|                                      | SSR HTML | SSR JSON | DSG HTML | DSG JSON | Static HTML | Static JSON | App Data | sw.js | Function Route | Generated Static Assets (Images, CSS, JS) /w Hash | Generated Static Assets (Images, CSS, JS) /wo Hash | Static Folder Assets | Gatsby Image CDN  |
|--------------------------------------|----------|----------|----------|----------|-------------|-------------|----------|-------|----------------|---------------------------------------------------|----------------------------------------------------|----------------------|-------------------|
| `getServerData` headers              |     Y    |     Y    |     N    |     N    |      N      |      N      |     N    |   N   |        N       |                         N                         |                          N                         |           N          |         N         |
| Gatsby Functions `setHeaders`        |     N    |     N    |     N    |     N    |      N      |      N      |     N    |   N   |        Y       |                         N                         |                          N                         |           N          |         N         |
| Default Immutable Cache Headers      |     N    |     N    |     N    |     N    |      N      |      N      |     N    |   N   |        N       |                         Y                         |                          N                         |           N          |         ??        |
| Default Revalidate Cache Headers     |     Y    |     Y    |     Y    |     Y    |      Y      |      Y      |     Y    |   Y   |        N       |                         N                         |                          Y                         |           Y          |         ??        |
| Default Security Headers             |    Y?    |    Y?    |    Y?    |    Y?    |      Y?     |      Y?     |    Y?    |   Y?  |        N       |                         Y?                        |                         Y?                         |          Y?          |         Y?        |
| gatsby-plugin-fastify Custom Headers |     Y    |     Y    |     Y    |     Y    |      Y      |      Y      |     Y    |   Y   |        N       |                         ??                        |                         ??                         |           Y          |         ??        |

The only thing not reflected here is which override which...I think that can just be explained.

moonmeister commented on integration-tests/plugin-fastify/src/api/test.js:
```diff
@@ -268,5 +270,9 @@ const responseData = [
 
 export default function handler(req, res) {
-  res.code(200).send(responseData);
+  if (isGatsbyHosting(res)) {
+    res.status(200).send(responseData);
+  } else {
+    res.code(200).send(responseData);
+  }
```
Been meaning to do this...thanks.

tsdexter commented:
@moonmeister  nothing to review tonight, but I made good progress today...will have a push ready sometime tomorrow which I think will definitely get us in a ready to release state. At least ready for a pre-release and more extensive testing if you want to do that first - our site is ~7000 pages and uses most features so I can do some testing there.

tsdexter commented:
> @moonmeister  nothing to review tonight, but I made good progress today...will have a push ready sometime tomorrow which I think will definitely get us in a ready to release state. At least ready for a pre-release and more extensive testing if you want to do that first - our site is ~7000 pages and uses most features so I can do some testing there. 

@moonmeister just finishing up the rest of the tests against your feature table and new benchmarks...gotta head out, but I'll try to get back to finishing it up and committing later tonight

tsdexter commented:
@moonmeister the test site with the updated version should build on gatsby cloud from the latest commit and I'll also get the `plugin-fastify` test site built and pushed to WPE Atlas platform running fastify on Monday so we can compare the two.

tsdexter commented:
- [ ] TODO: update docs to match refactor

tsdexter commented:
@moonmeister didn't get much time to work on this today...thoughts on https://github.com/gatsby-uc/plugins/pull/392#issuecomment-1474398906 before I get back to it tomorrow?

moonmeister commented:
Definitely seems like we're on the right track.

moonmeister commented:
RE: Benchmarks - Glad we did some testing to validate this is the correct path forward. 

RE: Table - Thanks for validating and giving feedback. I agree with your assessments. Only one question, re: Functions and Security headers, What does Gatsby Cloud do? 

RE: Refactor - Thanks, I'll do a code review and provide feedback there

moonmeister commented on packages/gatsby-plugin-fastify/src/gatsby-node.ts:
```diff
@@ -7,23 +8,60 @@ import type { GatsbyNode } from "gatsby";
+          mode: "FUNCTION",
+        })
+      ),
+      ...[...pluginData.pages].map(([, value]) => ({ path: value.path, mode: value.mode })),
```
What's up with this double expansion of the array?

tsdexter replied:
ESLint did that during `lint-staged`. It was originally `...Array.from(pluginData.pages).map`. `pluginData.pages` is of type `Map` which when you expand it returns an array of arrays with `[key, value]`: 

```js
[
  [0: "/path", 1: { mode: "SSG", path: "/path" }]
]
```

which we then `.map` over to get the array of objects we want and then we expand that into pages with the rest of them.

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/decorators.ts:
```diff
@@ -3,15 +3,22 @@ import fp from "fastify-plugin";
   interface FastifyReply {
     appendModuleHeader: typeof moduleHeaderDecorator;
     setHeader: FastifyReply["header"];
+    mode: Page["mode"] | "";
```
Would this be the same/better?

```suggestion
    mode: Page["mode"] | string;
```

tsdexter replied:
I don't think so. That would allow any string value vs the actual `modes` + `""`. Ideally, it would just be `Page["mode"]` - the only reason I added `| ""` was because Fastify docs recommend instantiating decorators with the same type they are intended to use, ie: `""` vs `null | undefined`.

moonmeister replied:
ah, right. That makes sense.

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/functions.ts:
```diff
@@ -49,6 +49,8 @@ export const handleFunctions: FastifyPluginAsync<{
               try {
                 reply.appendModuleHeader("Functions");
+                reply.mode = "FUNCTION";
+                reply.path = `${FUNCTIONS_PREFIX}${path}`;
```
Should we stay consistent with our use of `path.join`, or the POSIX version, which ever is correct?

tsdexter replied:
we can switch it to POSIX - just `path.join` does `\\api\\test` (on Windows).

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,46 @@
+    // add configured headers
+    reply.headers(pageHeaders);
+
+    // check pageHeaders for "undefined" values and remove those headers
```
I agree we shouldn't be sending undefined headers...this seems like a type checking issue somewhere else and should not need to be cleaned at runtime.

tsdexter replied:
This isn't a type checking issue. 

Perhaps you can add some input on a better way to handle this but in order to support being able to remove headers I made it so if you pass the `string` value `"undefined"` to the config object (or any of the default constants) it will properly remove the header with `reply.removeHeader`.

I originally needed this functionality to match gatsby hosting as for some routes they do not return a `cache-control` header at all, whereas Fastify was defaulting to `public, max-age: 0` if we didn't explicitly set or remove it. Afterwards, I realized it could be helpful for `customHeaders` as well to "unset" headers on specific paths and documented it in the README.

moonmeister replied:
Ah...okay. hmmm. I don't love the `"undefined"` but maybe there is no other good option. what about just using js `undefined`? or `false`? or `"unset"`?

moonmeister replied:
Also, I think we can do this in the `onPostBuild`...The only exception would be default security. but that could be cleaned up to be faster probably. Thoughts?

tsdexter replied:
> Ah...okay. hmmm. I don't love the `"undefined"` but maybe there is no other good option. what about just using js `undefined`? or `false`? or `"unset"`?

IIRC, setting the object like `{ "cache-control": undefined }` with the js `undefined` makes the object not have the `cache-control` property (it is undefined) so we can't check for it and remove the headers. I went back and forth with `"undefined"` and `"unset"` and ended on using undefined because the end result is that the header is not defined (vs the default or previously matched path headers). I think unset works too, it's just a more specific/made up term for our use-case. If you want to change it, I would lean back toward unset. 

> Also, I think we can do this in the onPostBuild...The only exception would be default security. but that could be cleaned up to be faster probably. Thoughts?

`onPostBuild` doesn't actually set or remove the headers, just puts the entry into the `headers` object, we could check previously configured defaults/customs to remove them here instead of adding the unset/undefined value - but we still have to check and `removeHeader` at runtime to cover cases where it's getting set by fastify outside of our own code, such as the `cache-control` default or other headers the user may want to remove. These lines are only looking at the specific headers object for the matched path, which is only 5-10 entries, generally, so I don't think it will make a performance difference to do some in `onPostBuild` if we still have to check/remove at runtime anyway. Thoughts?

moonmeister replied:
So, I'm guessing that `@fastify/static` is adding this header, see: https://github.com/fastify/fastify-static#send-options. If we disable that then we should not have to deal with any headers set by Fastify. There are a couple of others like `Etag` and `LastModified` but I don't see any reason these would need to be overridden. If someone brings one to light, we can address the issue then. This means the only possible things we'd need to undefine would be things the user is setting. or gatsby defaults...which is kinda why we have those as selectable options. Unless you have an immediate use-case for this I think we should shelve this as a feature. Wait till there's a clear need.

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,46 @@
+  useDefaultSecurity: boolean;
+}> = fp(async (fastify, { headers, useDefaultSecurity }) => {
+  fastify.addHook("onSend", async function (request, reply: FastifyReply) {
+    let pageHeaders = {};
```
Instead of mutating this, I think [`Object.assign` ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign) can clean this up.

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/server-routes.ts:
```diff
@@ -60,16 +62,10 @@ export const handleServerRoutes: FastifyPluginAsync<{
-            if (pageQueryData?.serverDataStatus) {
-              reply.code(pageQueryData.serverDataStatus);
-            }
+            if (pageQueryData?.serverDataStatus) reply.code(pageQueryData.serverDataStatus);
```
Did a formatter change these or was that you? I'd prefer to leave the `{  }` on if statements.

moonmeister replied:
Cool, yeah I started to do this for a bit and realized it was always a pain. 🤣

moonmeister reviewed (changes requested):
I don't think I have the literal bandwidth to make a release tonight...I'll try in the next couple of days when I have time and a decent cell conection.

moonmeister commented on integration-tests/plugin-fastify/benchmark.js:
```diff
@@ -1,11 +1,12 @@
   getConfig,
-} = require("gatsby-plugin-fastify/utils/config");
-const { serveGatsby } = require("gatsby-plugin-fastify/plugins/gatsby");
+} = require("gatsby-plugin-fastify/dist/utils/config");
+const { serveGatsby } = require("gatsby-plugin-fastify/dist/plugins/gatsby");
```
weird...it kinda makes sense...but why has it been working is just as strange.

moonmeister replied:
ya know what...this isn't automated...it probably hasn't been run since I made this change.

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,46 @@
+    if (urlHeaders) pageHeaders = { ...pageHeaders, ...urlHeaders };
+
+    // handle special path type routes
+    if (["CSR", "SSR", "REDIRECT", "FUNCTION", "PROXY", "404", "500"].includes(reply.mode)) {
```
So I thought about something after I went to bed. If you look at the `x-gatsby-server` headers that get appended by every module some routes get multiple modules appended. e.g. 404 and 500 routes hit those handlers but because they use the static file server the static handler is hit last. Have you confirmed you will ever actually get a 400 or 500? Some CSR might be similar.

tsdexter replied:
Yes, `404`, `500`, and `CSR`, set `reply.mode` and `reply.path` at the same time they append the module header, ie: 

```js
    reply.appendModuleHeader("404");
    reply.mode = "404";
    reply.path = "/404/";
```

So they look for the configured headers for their path, ie: `/404/`, even though the actual URL might be something like: `/some/non-existing/page`. That being said, the user could still configure `customHeaders` for `/some/non-existing/page` with a matching pattern entry if they needed to because `urlHeaders` are also added to `pageHeaders` alongside these `pathHeaders`.

One thing I have on my to do list to check is that I probably should be adding `pathHeaders` before `urlHeaders` because `pathHeaders` would generally just contain the defaults for those modules and `urlHeaders` would contain `customHeaders` that should be able to overwrite the defaults. I'll check up on this and add a test case.

moonmeister replied:
Not sure if you undestood what I was saying `x-gatsby-server` can return something like `SSR; 404; Static;`. Cause the SSR module tried to route the url, but then 404 ed, which serves the 404 file via the Static handler. If each module is overwriting the last, your `reply.mode` would be `static` not `404` when it arrives at the headers plugin.

tsdexter replied:
Thinking on this a little more, I think your concern is that the `static` module will overwrite the `reply.mode`. That is not the case because the `reply.mode` decorator is only set on modules that use dynamic routes. Static does not set it. There are two reasons for this, one is because we only need it for dynamic routes, additionally, although it could be useful to know the `mode` even for static/other routes there is some background why I didn't/couldn't set it, at least for static, [in this comment](https://github.com/gatsby-uc/plugins/pull/392#discussion_r1136091588):

>In the first refactor I did go down the path of tagging routes with their module by adding a decorator reply.module which I was going to then use in the headers plugin to set the FG_MODULE_HEADER based on the module used as well. I had this working for all modules except static.ts because it is globally scoped, since the fastify-static plugin doesn't have any hooks to run when it's actually serving a file the Static module was added to all replies, not just the ones actually served by static.

moonmeister commented on packages/gatsby-plugin-fastify/src/plugins/headers.ts:
```diff
@@ -0,0 +1,46 @@
+export const handleHeaders: FastifyPluginAsync<{
+  headers: Headers;
+  useDefaultSecurity: boolean;
+}> = fp(async (fastify, { headers, useDefaultSecurity }) => {
```
Also thought about logging...not a rush but before we merge we'll want a couple of debug logs noting the use of default security headers, default caching headers, and custom headers. custom headers can probably just say hey there are custom headers or aren't. see https://github.com/gatsby-uc/plugins/blob/main/packages/gatsby-plugin-fastify/src/plugins/client-routes.ts#L16 for an example.

tsdexter commented:
@moonmeister haven't forgot about this, just been having some other production issues and deadlines. Will hopefully get back to it later this week or next.

tsdexter commented:
> @tsdexter You're not waiting on me are you?

Nope, no worries. Just had to get some other stuff done in the meantime. See above.

moonmeister commented:
Gatsby officially has an Adapter spec on the way 🥳 . It specifically notes the inclusion of headers in the spec to eliminate our very need to do this crazy implementation per adapter. 🥳 

https://github.com/gatsbyjs/gatsby/discussions/38231

tsdexter commented:
> Gatsby officially has an Adapter spec on the way 🥳 . It specifically notes the inclusion of headers in the spec to eliminate our very need to do this crazy implementation per adapter. 🥳
> 
> [gatsbyjs/gatsby#38231](https://github.com/gatsbyjs/gatsby/discussions/38231)

And I had it on the schedule to finally get back to this after my vacation in July. I wonder how long until it's stable - looks great so far, been playing around with the demo repo.

moonmeister commented:
> > Gatsby officially has an Adapter spec on the way 🥳 . It specifically notes the inclusion of headers in the spec to eliminate our very need to do this crazy implementation per adapter. 🥳
> > 
> > [gatsbyjs/gatsby#38231](https://github.com/gatsbyjs/gatsby/discussions/38231)
> 
> And I had it on the schedule to finally get back to this after my vacation in July. I wonder how long until it's stable - looks great so far, been playing around with the demo repo. 

Yeah but also this is going to make life so easy. We can close every issue ever and it'll just work...hopefully.

moonmeister commented:
With the announcement of the [Gatsby Adapters](https://github.com/gatsbyjs/gatsby/discussions/38231) spec, we'll be discontinuing any work on this plugin. Should the need for a performant node server still be needed, this work can continue under the official spec as `gatsby-adapter-fastify`.

tsdexter commented:
@moonmeister this makes sense. Just one comment. I think there is still a bit of uncertainty around adapters, but to me it seems they are intended to be more platform specific.

I think there is probably still a path for something generic like `gatsby-adapter-fastify` or `gatsby-adapter-express` etc but from the initial post in the RFC I think it may be intended to be less generic and more specific to the platform. 

For example, `gatsby-adapter-atlas` or `gatsby-adapter-google-cloud` either of which may still use `fastify` or `express` under the hood, but also takes care of the specific requirements for deployment on their respective platforms. 

Thoughts?

moonmeister commented:
@tsdexter I just opened #440 , let's have this discussion there...thanks for the feedback, I'll respond there.
